### ISSUE TITLE
Proposal for safe async I/O with `ArrayBuffer`s

### ISSUE THREAD

**Issue by andreubotella** [denoland/deno#16756]
In Rust, it is undefined behavior if a memory slice (which doesn't contain types with interior mutability) is mutated while it is borrowed, other than through the borrow for mutable borrows. In Deno's async ops, if a `ZeroCopyBuf` is deref'd across an await point, this is unsound, because the corresponding `ArrayBuffer` might be reachable from JS and modified in that await point. There are solutions that don't have this problem, like using `DetachedBuffer` (see #16722 for my previous proposal that used that), but that is not compatible with all of Deno's existing APIs, and the copying needed to make it compatible might hurt performance.

The main issue is that, for most of the I/O that Deno does, the APIs provided by Tokio are async and take `&[u8]` or `&mut [u8]`, which therefore stay borrowed for the duration of the async operation. But (for what I understand), these APIs are actually implemented as series of synchronous reads and writes behind the scenes:

```rust
// Pseudocode
async fn read(&self, buf: &mut [u8]) {
  let mut idx = 0;
  loop {
    let (read, status) = self.inner_read(&mut buf[idx..]);
    idx += read;
    match status {
      Done => return,
      NeedsToWait(future) => future.await,
      Continue => {},
    }
  }
}
```

So what if we only borrow the buffer's memory when performing a synchronous read, dropping the borrow before waiting? This would solve the unsoundness without having to detach buffers or make copies.

This, however, is quite low-level, at the level of `mio` rather than tokio. But `deno_core`'s `Buf(Mut)View` is a view over a `ZeroCopyBuf` that includes a cursor, and it could be used with this. Furthermore, my understanding is that @bnoordhuis is currently working on a new event loop for Deno that sits on top of `mio`, so maybe this could be built into that event loop, offering simple async functions that take a `Buf(Mut)View` as a parameter.

However, filesystem accesses don't work like this, and in fact Deno, as well as Tokio and `async-std` all achieve asynchronous filesystem operations by spawning a blocking task in a separate thread. From what I gather, the reason for this is that `std`'s I/O APIs ([`Read`](https://doc.rust-lang.org/std/io/trait.Read.html) and [`Write`](https://doc.rust-lang.org/std/io/trait.Write.html)) freely allow implementers to block the current thread, and they offer no way to detect whether an operation would block, or to notify when it would no longer block. But in Unix systems, open files are represented by file descriptors, which can be used with `mio`'s [`SourceFd`](https://docs.rs/mio/latest/mio/unix/struct.SourceFd.html) just like other I/O operations. And for Windows, my (admittedly very sparse) understanding is that you can open a file in non-blocking mode, so `Read` and `Write` are guaranteed not to block. So something like this might be doable with files after all.

**Comment by andreubotella** [comment]
I just realized that we'd have to figure out how to make this work with the [TC39 resizable buffers proposal](https://github.com/tc39/proposal-resizablearraybuffer), and it might not be trivial. But doing async I/O in repeated non-blocking synchronous steps _can_ work with a resizing buffer, which is much better than borrowing a slice that then might end up covering memory past the end of the `ArrayBuffer` (as it would be in the way Deno does I/O currently).

**Comment by ry** [comment]
I don't think this is a problem worth solving. I wouldn't call this behavior a problem nor would I call it unsound.

> So what if we only borrow the buffer's memory when performing a synchronous read, and drop the borrow before waiting? This would solve the unsoundness without having to detach buffers or make copies.

How does this improve the experience for the user? The user still allocates a buffer in JS, passes it to rust, some time passes and the buffer is modified, the notification of that change comes when the promise resolves.

**Comment by andreubotella** [comment]
> How does this improve the experience for the user? The user still allocates a buffer in JS, passes it to rust, some time passes and the buffer is modified, the notification of that change comes when the promise resolves.

The whole point of this proposal, as opposed to other ways to solve this unsoundness, is that it does not break any API or decrease performance. The point is to not change the experience for the user, since alternative solutions (such as #16722, for example) would worsen the user's experience.

---

> I don't think this is a problem worth solving. I wouldn't call this behavior a problem nor would I call it unsound.

As far as I'm concerned, undefined behavior is always a bug. Per the Rust reference, undefined behavior in Rust [includes](https://doc.rust-lang.org/reference/behavior-considered-undefined.html):

> - Breaking the [pointer aliasing rules](http://llvm.org/docs/LangRef.html#pointer-aliasing-rules). `&mut T` and `&T` follow LLVM’s scoped [`noalias`](http://llvm.org/docs/LangRef.html#noalias) model, except if the `&T` contains an [`UnsafeCell<U>`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html).
> - Mutating immutable data. All data inside a [`const`](https://doc.rust-lang.org/reference/items/constant-items.html) item is immutable. Moreover, all data reached through a shared reference or data owned by an immutable binding is immutable, unless that data is contained within an [`UnsafeCell<U>`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html).

My understanding is that these bullet points mean that the memory pointed to by a `&[u8]` can never be mutated while the `&[u8]` is alive, and the memory pointed to by a `&mut [u8]` (which is used to mutate the slice) can never be accessed (mutably or immutably) while the `&mut [u8]` is alive, except through the very same `&mut [u8]`.

Now consider this code:

```ts
const conn: Deno.TcpConn = await Deno.connect({ port: 5421 });

const buffer: Uint8Array = new TextEncoder().encode("Test");
const writePromise: Promise<number> = conn.write(buffer);  // We don't await!
queueMicrotask(() => {
  buffer[0] = 127;
});
```

`Deno.TcpConn`'s implementation of the `Writer` interface reads on a `TcpStreamResource`, which seems to borrow the memory for the passed `Uint8Array` synchronously – before the microtask runs. This can be tested by adding a `println!` on `impl_writable` before the first await, and adding a `console.log` on the microtask. Therefore, if the write can't happen synchronously (which is unlikely for such a small buffer, but way more likely for larger buffers), the buffer's memory contents will be borrowed as a `&[u8]` while JS is writing into the same memory. Therefore, you get undefined behavior.

**Comment by littledivy** [comment]
The buffer model wasn't fundamentally changed since this was filed: `ZeroCopyBuf` was renamed to `JsBuffer` / `V8Slice`, which still holds a `v8::SharedRef<v8::BackingStore>` plus a range and derefs into the live, JS-reachable backing store (`libs/serde_v8/magic/v8slice.rs`). Async I/O ops still take this buffer (`#[buffer] mut buf: JsBuffer`) and hold it across await points, so the soundness concern described here still applies structurally — it was not resolved by a code change. (`V8Slice::as_slice` does now clamp to the current backing-store length, which partially addresses the resizable-buffer point, but not the core borrow-across-await UB.)
